src/main/scala/mesosphere/marathon/api/v2/TasksResource.scala

```diff
@@ -41,50 +42,58 @@ class TasksResource @Inject() (
   @GET
   @Produces(Array(MarathonMediaType.PREFERRED_APPLICATION_JSON))
   @Timed
+  @SuppressWarnings(Array("all")) /* async/await */
   def indexJson(
     @QueryParam("status") status: String,
     @QueryParam("status[]") statuses: util.List[String],
     @Context req: HttpServletRequest): Response = authenticated(req) { implicit identity =>
     Option(status).map(statuses.add)
     val conditionSet: Set[Condition] = statuses.flatMap(toTaskState)(collection.breakOut)
 
-    val instancesBySpec = instanceTracker.instancesBySpecSync
-
-    val instances = instancesBySpec.instancesMap.values.view.flatMap { appTasks =>
-      appTasks.instances.map(i => appTasks.specId -> i)
-    }
-
-    val appIds = instancesBySpec.allSpecIdsWithInstances
-
-    val appIdsToApps: Map[PathId, Option[AppDefinition]] =
-      appIds.map(appId => appId -> result(groupManager.app(appId)))(collection.breakOut)
-
-    val appToPorts = appIdsToApps.map {
-      case (appId, app) => appId -> app.map(_.servicePorts).getOrElse(Nil)
-    }
-
-    val health: Map[Instance.Id, Seq[Health]] = appIds.flatMap { appId =>
-      result(healthCheckManager.statuses(appId))
-    }(collection.breakOut)
-
-    val enrichedTasks: Iterable[EnrichedTask] = instances.flatMap {
-      case (appId, instance) =>
-        val app = appIdsToApps(appId)
-        if (isAuthorized(ViewRunSpec, app) && (conditionSet.isEmpty || conditionSet(instance.state.condition))) {
-          instance.tasksMap.values.map { task =>
-            EnrichedTask(
-              appId,
-              task,
-              instance.agentInfo,
-              health.getOrElse(instance.instanceId, Nil),
-              appToPorts.getOrElse(appId, Nil)
-            )
+    val futureEnrichedTasks = async {
+      val instancesBySpec = await(instanceTracker.instancesBySpec)
+
+      val instances = instancesBySpec.instancesMap.values.view.flatMap { appTasks =>
+        appTasks.instances.map(i => appTasks.specId -> i)
+      }
+      val appIds = instancesBySpec.allSpecIdsWithInstances
+
+      val appIdsToApps: Map[PathId, Option[AppDefinition]] = await(
+        Future.sequence(
+          appIds.map(appId => groupManager.app(appId).map(
```

Comment:
This should work, right?

```scala
val appIdsToApps: Map[PathId, Option[AppDefinition]] =
  await(Future.sequence(appIds.map(appId => groupManager.app(appId).map(appId -> _ ))).to[Map]
```

It's still messy. Ideally `GroupManager` has a method to retrieve this map. Hm.